nemo/backends/pytorch/actions.py

```diff
@@ -1418,26 +1399,22 @@ def train(
                 curr_tensors_to_optimize = training_loop[self.step % len(training_loop)][1]
                 final_loss = 0
                 for tensor in curr_tensors_to_optimize:
-                    if (
-                        torch.isnan(registered_tensors[tensor.unique_name]).any()
-                        or torch.isinf(registered_tensors[tensor.unique_name]).any()
-                    ):
-                        if (
-                            (stop_on_nan_loss)
-                            or (self._optim_level not in AmpOptimizations)
-                            or (self._optim_level == Optimization.mxprO0)
-                        ):
-                            # Set flag here and terminate at next all_gather check.
-                            nan_or_inf = True
-                            logging.warning(
-                                'Loss is NaN or inf at step %d, will terminate within the'
-                                ' next steps_per_nan_check steps',
-                                self.step,
-                            )
-                        else:
-                            logging.warning('Loss is NaN or inf, continuing training')
                     final_loss += registered_tensors[tensor.unique_name]
 
+                # Check for NaN/inf loss (across workers if applicable)
+                loss_nan_inf_checker = final_loss.clone()
+                if placement_gpu:
+                    dist.all_reduce(loss_nan_inf_checker)
```

Comment:
Isn't an overflow possible here? Since the default is SUM.

What is the reasoning to check every step?

Reply:
Good catch, apparently I forgot that floats overflow to inf. Changed this to use MAX.

We should check every step to prevent NaN/inf loss from ever propagating any garbage gradient at any step.